lib/puppet/transaction.rb

```diff
@@ -227,30 +228,21 @@ def failed?(resource)
 
   # Does this resource have any failed dependencies?
   def failed_dependencies?(resource)
-    # First make sure there are no failed dependencies.  To do this,
-    # we check for failures in any of the vertexes above us.  It's not
-    # enough to check the immediate dependencies, which is why we use
-    # a tree from the reversed graph.
-    found_failed = false
-
-
-    # When we introduced the :whit into the graph, to reduce the combinatorial
-    # explosion of edges, we also ended up reporting failures for containers
-    # like class and stage.  This is undesirable; while just skipping the
-    # output isn't perfect, it is RC-safe. --daniel 2011-06-07
-    suppress_report = (resource.class == Puppet::Type.type(:whit))
-
-    relationship_graph.dependencies(resource).each do |dep|
-      next unless failed?(dep)
-      found_failed = true
-
-      # See above. --daniel 2011-06-06
-      unless suppress_report then
-        resource.notice "Dependency #{dep} has failures: #{resource_status(dep).failed}"
+    s = resource_status(resource) and s.dependency_failed?
+  end
+
+  # We need to know if a resource has any failed dependencies before
+  # we try to process it. We keep track of this by keeping a
+  # `dependency_failed` flag on each resource and incrementally
+  # computing it as the OR of the flag of each dependency. We have to
+  # do this as-we-go instead of up-front at failure time because the
+  # graph may be mutated as we walk it.
+  def propagate_failure(resource)
+    relationship_graph.direct_dependencies_of(resource).each do |dep|
+      if (s = resource_status(dep)) and (s.failed? || s.dependency_failed?)
+        resource_status(resource).dependency_failed = true
```

Comment:
Since we're only looking at direct dependencies (parents), I think we should add a test that verifies grandchildren are flagged as failed correctly. For example, if we have `A->B->C`, and A fails, then when we `eval_resource(B)`, it will see that its parent A has failed, and mark B as `dependency_failed = true`. Later we expect to do `eval_resource(C)` and expect that C will be marked as `dependency_failed = true` as well (since B has failed too).